framework/include/utils/FactorizedRankTwoTensor.h

```diff
@@ -0,0 +1,291 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "RankTwoTensor.h"
+#include "SymmetricRankTwoTensor.h"
+
+// forward declarations
+template <typename>
+class FactorizedRankTwoTensorTempl;
+
+/**
+ * FactorizedRankTwoTensorTempl is designed to perform the spectral decomposition of an
+ * underlying symmetric second order tensor and reuse its bases for future operations if possible.
+ *
+ * FactorizedRankTwoTensorTempl templates on the underlying second order tensor.
+ * IMPORTANT: the underlying second order tensor must be symmetric. A check is only performed in
+ * debug mode.
+ *
+ * Only operations that reuses the known factorization are provided. Otherwise, you will need to
+ * first retrieve the underlying RankTwoTensorTempl to perform the operation.
+ *
+ * TODO? Although I am calling it factorization, it only really refers to eigenvalue decompositon at
+ * this point. I am not sure if in the future we need to add other types of similarity
+ * transformation.
+ */
+template <typename T>
+class FactorizedRankTwoTensorTempl
+{
+public:
+  /// For generic programming
+  typedef typename T::value_type value_type;
+
+  /// No default constructor
+  FactorizedRankTwoTensorTempl() = delete;
+
+  /// Copy constructor
+  FactorizedRankTwoTensorTempl(const FactorizedRankTwoTensorTempl<T> & A) = default;
+
+  /// Constructor if the factorization isn't known a priori
+  FactorizedRankTwoTensorTempl(const T & A)
+  {
+    if (!A.isSymmetric())
+      mooseError("The tensor is not symmetric.");
+    A.symmetricEigenvaluesEigenvectors(_eigvals, _eigvecs);
+  }
+
+  // Construct from the factorization
+  // Assume that regardless of the type of T, the eigenvectors are always stored as as a full second
+  // order tensor, e.g. the eigenvector matrix of a symmetric second order tensor isn't symmetric in
+  // general. Hence we don't take advantage of orthonormal and/or unitary second order tensors.
+  FactorizedRankTwoTensorTempl(const std::vector<typename T::value_type> & eigvals,
+                               const RankTwoTensorTempl<typename T::value_type> & eigvecs)
+    : _eigvals(eigvals), _eigvecs(eigvecs)
+  {
+  }
+
+  // @{ getters
+  template <typename T2 = T>
+  T2 get() const
+  {
+    return static_cast<T2>(assemble());
+  }
+  template <typename T2 = T>
+  T2 get()
+  {
+    return static_cast<T2>(assemble());
+  }
+  const std::vector<typename T::value_type> & eigvals() const { return _eigvals; }
+  std::vector<typename T::value_type> eigvals() { return _eigvals; }
```

Comment:
Uhm, is that non const getter even needed? why would we get them by copy in that case? why not also get them by const reference?